From 8ff33f92ed012fc67ff9c9e0af3858034d4dd87f Mon Sep 17 00:00:00 2001
From: Christian Ehrlicher <ch.ehrlicher@gmx.de>
Date: Tue, 29 Sep 2026 17:58:47 +0200
Subject: [PATCH] QCommonStyle: revert using QIcon::paint() for drawing pixmaps

Revert 226b2e8508ca6b6a9cbf2c5fcc12d550338a5e28 as QIcon::paint() is
using the target rect in device independent coordinates which leads to
rouding errors and in the end to wrong (rounded) target rect
coordinates.
The same problem can be found when using QPainter::drawPixmap(QRect)
when drawing on a scaled device.

Fixes: QTBUG-148522
Pick-to: 6.11
Change-Id: I4dfb43114f094428d7b991743cb7da4c269aae0c
Reviewed-by: Axel Spoerl <axel.spoerl@qt.io>
(cherry picked from commit d8c6f378f9e9fa253c512fa40a0a0395b200e54d)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
Origin: upstream, https://invent.kde.org/qt/qt/qtbase/-/commit/8ff33f92ed012fc67ff9c9e0af3858034d4dd87f
Forwarded: not-needed
Last-Update: 2026-10-29
---
 src/widgets/styles/qcommonstyle.cpp | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/widgets/styles/qcommonstyle.cpp b/src/widgets/styles/qcommonstyle.cpp
index 5bfaea67ae6..142b3e7407a 100644
--- a/src/widgets/styles/qcommonstyle.cpp
+++ b/src/widgets/styles/qcommonstyle.cpp
@@ -1375,7 +1375,7 @@ void QCommonStyle::drawControl(ControlElement element, const QStyleOption *opt,
                 if (button->state & (State_On | State_Sunken))
                     iconRect.translate(proxy()->pixelMetric(PM_ButtonShiftHorizontal, opt, widget),
                                        proxy()->pixelMetric(PM_ButtonShiftVertical, opt, widget));
-                p->drawPixmap(iconRect, pixmap);
+                p->drawPixmap(iconRect.x(), iconRect.y(), pixmap);
             } else {
                 tf |= Qt::AlignHCenter;
             }
@@ -1981,7 +1981,9 @@ void QCommonStyle::drawControl(ControlElement element, const QStyleOption *opt,
             if (!tab->icon.isNull()) {
                 const auto mode = (tab->state & State_Enabled) ? QIcon::Normal : QIcon::Disabled;
                 const auto state = (tab->state & State_Selected) ? QIcon::On : QIcon::Off;
-                tab->icon.paint(p, iconRect, Qt::AlignCenter, mode, state);
+                const auto tabIcon =
+                        tab->icon.pixmap(tab->iconSize, QStyleHelper::getDpr(p), mode, state);
+                p->drawPixmap(iconRect.x(), iconRect.y(), tabIcon);
             }
 
             proxy()->drawItemText(p, tr, alignment, tab->palette, tab->state & State_Enabled, tab->text,
